### PR TITLE
Adds donation api

### PR DESCRIPTION
**Pull request by oreHGA**
This PR adds functionality to:
- Create new donation upon request from client
- Validate payment status of a given donation (update&return records)

@xvd3  please take a look and let me know what you think

### CODE REVIEW

File: services/donations/validate.js [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+ * Formats donation record before sending to client
+ * 
+ * @param {*} donation 
+ */
+async function clientDonationRecord( donation ) {
+
+  // call a ben ids to mongoose objects
+  let beneficiary_ids = donation.beneficiary_ids;
+  beneficiary_ids = beneficiary_ids.map(id => {
+    return mongoose.Types.ObjectId("5e9b78595d935a2468c74558")
```
Comment (xvd3) [inline services/donations/validate.js:138]: Test id here right?

File: services/donations/validate.js [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+    if( donation === null ) {
+      throw new Error( "Donation not found" );
+    }
+
+    // if payment is in initiated state, attempt to update else return
+    if( donation.status !== "INITIATED" ) {
+      d.resolve( clientDonationRecord(donation) );
+    } else {
+      // validate payment from provider & update records
+      validate_payment(donation).then(async (res) => {
```
Comment (xvd3) [inline services/donations/validate.js:41]: Any reason you are not using the await sync flow here? 
Something like this https://github.com/vunderkind/gf-backend/pull/9/files#diff-cec8dc8aa495e3da64081369e27369c7R42

File: services/donations/validate.js [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+    headers: { "Content-Type": "application/json" }
+  }
+  let body = {
+    "SECKEY": process.env.FLUTTERWAVE_SEC_KEY,
+    "txref": donation.reference
+  }
+
+  let result = {};
+
+  await axios.post(process.env.FLW_VERIFY_ENDPOINT, body, config)
```
Comment (xvd3) [inline services/donations/validate.js:88]: Same here. I think the await is redundant. Could use this style https://github.com/vunderkind/gf-backend/pull/9/files#diff-cec8dc8aa495e3da64081369e27369c7R42
 Or use the axios promise directly

File: services/donations/validate.js [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+  .then(function(res) {
+    const payment_info = res.data.data;
+
+    result = {
+      "code": res.status
+    };
+
+    if (payment_info.status === "successful" && payment_info.chargecode == 00) {
```
Comment (xvd3) [inline services/donations/validate.js:96]: Any reason for checking the chargecode as well?

File: services/donations/validate.js [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+          "chargedamount": payment_info.chargedamount,
+          "ip": payment_info.ip,
+          "fraudstatus": payment_info.fraudstatus,
+          "paymenttype": payment_info.paymenttype
+        })
+    }
+
+  }).catch(function (error) {
+    // Prepsre response
+    result = {
```
Comment (xvd3) [inline services/donations/validate.js:117]: What do you think about wrapping this bit with an if-block. Thinking for cases where error is not an instance of axios' response. doing error.response.data.message (when response is not part of the error object could cause that last block to fail)

File: services/donations/validate.js (hunk 4 above)
Comment (oreHGA) [reply services/donations/validate.js:96]: Noticed all the doc samples did. Decided to follow suit https://developer.flutterwave.com/docs/status-check

Comment (oreHGA) [reply services/donations/validate.js:96]: Do you have any recommendations against checking for the chargecode?

File: services/donations/validate.js (hunk 2 above)
Comment (oreHGA) [reply services/donations/validate.js:41]: The `validate_payment` method returns a promise and then I handle the success/failure using `then` and `catch`. I can swap the styles though to keep the codebase consistent with yours

Comment (xvd3) [reply services/donations/validate.js:41]: Oh ok, I just wanted to confirm

File: services/donations/validate.js (hunk 4 above)
Comment (xvd3) [reply services/donations/validate.js:96]: Nothing at all. The status field is a good enough source, for truth. Once it's successful, nothing else matters. (Checking 00's a part of the old style for validating)

Comment (oreHGA) [reply services/donations/validate.js:96]: Nice, will remove it then

(they need to update their docs 🌚)

File: services/donations/validate.js (hunk 5 above)
Comment (oreHGA) [reply services/donations/validate.js]: Thanks for pointing that out, handling non axios errors now

File: services/donations/validate.js (hunk 2 above)
Comment (oreHGA) [reply services/donations/validate.js]: Updated the flow now. Your method actually looks more readable

File: services/donations/create.js [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+    //Init defaults
+    params.created_at = new Date().toUTCString();
+    params.created_ts = Date.now();
+    params.status = "INITIATED";
+    params.reference = uuid.v4();
+
+    const bens_valid = await validate_beneficiaries(params.beneficiary_ids);
+    if ( !bens_valid ) {
+      d.reject({
+        message: "Invalid beneficiary_ids, please retry"
```
Comment (xvd3) [inline services/donations/create.js:33]: Hi, so here, some node versions might show warnings that reject is being called with something that's not an instance of the Error class. Could consider changing to:
```
if (!bens_valid) throw new Error("Invalid beneficiary_ids, please retry");
```

The reject in the catch block will handle the rest. Message passed to new Error will be the value for `e.message` in the route handler

File: services/donations/validate.js [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+        // this is to provide records on our side incase of disputes/fraud
+        result['gw_response'] = JSON.stringify({
+          "chargedamount": payment_info.chargedamount,
+          "ip": payment_info.ip,
+          "fraudstatus": payment_info.fraudstatus,
+          "paymenttype": payment_info.paymenttype
+        })
+    }
+  } catch (error) {
+    if( error.isAxiosError == true ) {
```
Comment (xvd3) [inline services/donations/validate.js:114]: `if( error.isAxiosError )`

File: services/donations/create.js (hunk 6 above)
Comment (oreHGA) [reply services/donations/create.js]: Thanks!

**Pull request merged**